commit e0097a39eeeb12615476c38ffea4634ed8089dac
Author: Vasily Postnicov <shamaz.mazum@gmail.com>
Date:   Sat Sep 26 19:55:03 2026 +0300

    Improve NODE-CONSERVATIVE-TYPE for conses.

    This means that you can elide type declarations for cons args
    if you already have a FTYPE declaration and aren't doing bad things
    to the arg. E.g.:

    (declaim (ftype (function ((cons fixnum fixnum)) (values t &optional)) foo))
    (defun foo (x)
      (declare (optimize (speed 3))
               ;; You can remove the following declaration and get this
               ;; function compiled without notes.
               #+nil
               (type (cons fixnum fixnum) x))
      (logand (car x) (cdr x)))

diff --git src/compiler/ir1opt.lisp.orig src/compiler/ir1opt.lisp
index cace54dcf..5c0eaa0c7 100644
--- src/compiler/ir1opt.lisp.orig
+++ src/compiler/ir1opt.lisp
@@ -241,11 +241,26 @@
            (let* ((derived-values-type (node-derived-type node))
                   (derived-type (single-value-type derived-values-type)))
              (if (ref-p node)
-                 (let ((leaf (ref-leaf node)))
-                   (if (and (basic-var-p leaf)
-                            (cdr (leaf-refs leaf)))
-                       (coerce-to-values (conservative-type derived-type))
-                       derived-values-type))
+                 (let ((leaf (ref-leaf node))
+                       (cons-type (specifier-type 'cons)))
+                   (if (or (not (basic-var-p leaf))
+                           (null (cdr (leaf-refs leaf)))
+                           (and
+                            ;; Works only for conses
+                            (csubtypep derived-type cons-type)
+                            ;; All functions for which this var is an arg are
+                            ;; pure.
+                            (every
+                             (lambda (ref)
+                               (let ((dest (node-principal-dest ref)))
+                                 (and (combination-p dest)
+                                      (flushable-combination-p dest))))
+                             (leaf-refs leaf))
+                            ;; This var is not assigned to and we have no
+                            ;; other dests.
+                            (null (basic-var-sets leaf))))
+                       derived-values-type
+                       (coerce-to-values (conservative-type derived-type))))
                  derived-values-type))))
     (if (listp node)
         (apply #'values-type-union (mapcar #'node-type node))
diff --git src/compiler/ir1util.lisp.orig src/compiler/ir1util.lisp
index fa8417d6f..d3c12ba96 100644
--- src/compiler/ir1util.lisp.orig
+++ src/compiler/ir1util.lisp
@@ -52,7 +52,16 @@
         (setf (block-last block) (ctran-use next))
         (setf (node-next (block-last block)) nil)
         block))))
-
+
+;; Find the node's dest bypassing casts.
+(defun node-principal-dest (node)
+  (declare (type node node))
+  (labels ((%go (node)
+             (let ((dest (node-dest node)))
+               (if (cast-p dest)
+                   (node-principal-dest dest)
+                   dest))))
+    (%go node)))
 ;;;; lvar use hacking
 
 ;;; Return a list of all the nodes which use LVAR.
diff --git tests/compiler-2.pure.lisp.orig tests/compiler-2.pure.lisp
index 84db62632..d162c7d14 100644
--- tests/compiler-2.pure.lisp.orig
+++ tests/compiler-2.pure.lisp
@@ -4730,13 +4730,22 @@
    (lambda ()
      (let ((x '(1 2 3)))
        (values (pop x) (pop x))))
-   (values (integer 1 3) (or null (integer 1 3)) &optional))
+   (values (integer 1 3) (integer 1 3) &optional))
   (assert-type
    (lambda ()
      (declare (optimize speed (debug 1)))
      (destructuring-bind (a b) '(1 2)
        (values a b)))
-   (values (integer 1 2) (or null (integer 1 2)) &optional))
+   (values (integer 1 2) (integer 1 2) &optional))
+  (assert-type
+   (lambda ()
+     (let ((x '(1 2)))
+       (values (pop x)
+               (pop x)
+               (pop x))))
+   (values (integer 1 2)
+           (integer 1 2)
+           (or (integer 1 2) null) &optional))
   (assert-type
    (lambda (n)
      (loop for x in '(a b c)
@@ -4791,6 +4800,39 @@
        0))
    (eql 0)))
 
+(with-test (:name :cons-conservative-type)
+  (macrolet ((do-check (type &body bindings)
+               `(assert-type
+                 (lambda (x)
+                   (let* (,@bindings)
+                     (declare (type (function ((cons fixnum fixnum))
+                                              (values t &optional))
+                                    f))
+                     (funcall f x)))
+                 ,type)))
+    (do-check fixnum
+      (f (lambda (cons)
+           (logand (car cons)
+                   (cdr cons)))))
+    (do-check integer
+      (foo (lambda (x) x))
+      (f (lambda (cons)
+           (funcall foo cons)
+           (logand (car cons)
+                   (cdr cons)))))
+    (do-check integer
+      (f (lambda (cons)
+           (let ((foo cons))
+             (declare (special foo))
+             (logand (car cons)
+                     (cdr cons))))))
+    (do-check integer
+      (f (lambda (cons)
+           (let ((foo cons))
+             (setf (car foo) (random (expt 2 100)))
+             (logand (car foo)
+                     (cdr foo))))))))
+
 (with-test (:name :test-headers-lowtag-only)
   (checked-compile-and-assert
    ()
